Firestore/core/src/firebase/firestore/remote/remote_store.h

```diff
@@ -177,7 +178,9 @@ class RemoteStore : public TargetMetadataProvider,
   void AddToWritePipeline(FSTMutationBatch* batch);
 
   /** Returns a new transaction backed by this remote store. */
-  FSTTransaction* CreateTransaction();
+  // TODO(c++14): return a plain value when it becomes possible to move
```

Comment:
Returning a `shared_ptr` is largely an optimization. The transaction is copied into a lambda more than once, and moving into a lamdba is not possible in C++11.

Reply:
sgtm

Reply:
Does each lambda has to retain an ownership of Transaction? I expect each lambda only use the passed in Transaction instead of retain an ownership of it. Whoever started a transaction would be the sole owner of the Transaction.

Reply:
The problem is the lambda in [`FSTSyncEngine.mm`](https://github.com/firebase/firebase-ios-sdk/pull/2362/files/7f47e36bcbd7eb01d2c3ca3fef7ed7223b2895a7#diff-5e4ee796c53a5b08cbf382de7b23fc73R297). It creates a `Transaction` as a local variable and immediately passes it into a lambda.